---
title: "ATP_Model_Performance_caret"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(caret)
library(pROC)
library(dplyr)

library(cvAUC)

library(cowplot)
library(randomForest)
library(ggbeeswarm)
library(reshape2)

col_list <- col_list <- c('#b74a8c',
              '#8c7029',
              '#7464c0',
              '#ac5a51')

make_pROC_curve_mean <- function(list_of_models, col_list, legend=F){
  
  text_pos <- c(0.3,0.25,0.2,0.15)
  add <- F
  
  roc_list <- list()
  for(i in 1:length(list_of_models)){
    name <- names(list_of_models)[i]
    mean_votes <- list_of_models[[i]]$pred %>% group_by(rowIndex) %>% summarize_if(is.numeric, list(name=mean))
    mean_votes$obs <- list_of_models[[i]]$trainingData$.outcome[mean_votes$rowIndex]
    
    temp_roc <- roc(response=mean_votes$obs, predictor = mean_votes$case_name, ci=T, levels=c("control", "case"),
                    direction="<")
    roc_list[[name]] <- temp_roc
  }
  ci.list <- lapply(roc_list, ci.se, specificities=seq(0,1,0.05))
  
  dat.ci.list <- lapply(ci.list, function(ciobj)
    data.frame(x=as.numeric(rownames(ciobj)),
               lower=ciobj[,1],
               upper=ciobj[,3]))
  
  dat.auc.list <- lapply(roc_list, function(x)
    data.frame(lower=x$ci[1],
               mid=x$ci[2],
               upper=x$ci[3]))
  
  p <- ggroc(roc_list) + theme_minimal() + 
   geom_abline(slope=1, intercept = 1, linetype = "dashed", alpha=0.7, color = "grey") +
   coord_equal() +
     scale_color_manual("Feature type", values=col_list) +
     scale_fill_manual("Feature type", values=col_list)
  


  
  for(i in 1:length(dat.ci.list)) {
    dat.ci.list[[i]]$name <- factor(names(dat.ci.list)[i], levels=c("ASV RA", "ASV CLR", "Genus CLR", "Genus RA"))
    
    p <- p + geom_ribbon(
      data = dat.ci.list[[i]],
      aes(x = x, ymin = lower, ymax = upper, fill=name),
      alpha = .1,
      inherit.aes = F)
  
    dat.auc.list[[i]]$name <- factor(names(dat.ci.list)[i], levels=c("ASV RA", "ASV CLR", "Genus CLR", "Genus RA"))
    

    p <- p + geom_text(
      data=dat.auc.list[[i]],
      aes(x=.25,
          label = paste("AUC: ", round(mid, 3), " (", round(lower, 3), " - ", round(upper, 3), ")", sep=""),
          color=name),
      inherit.aes=F,
      y=text_pos[i],
      show.legend=FALSE)
    message(text_pos[i])
  } 
  
  return(p)
}


```

# ROC models

## Breast

```{r}
Br_ASV_RA <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/Breast/ASV_RCV_RA_caret.RDS")
Br_ASV_CLR <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/Breast/ASV_RCV_CLR_caret.RDS")

Br_Genus_RA <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/Breast/Genus_RCV_RA_caret.RDS")
Br_Genus_CLR <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/Breast/Genus_RCV_CLR_caret.RDS")

Br_models <- list("ASV CLR"=Br_ASV_CLR, 
                  "ASV RA"=Br_ASV_RA, 
                  "Genus CLR"=Br_Genus_CLR, 
                  "Genus RA"=Br_Genus_RA)

Br_plot_mean <- make_pROC_curve_mean(Br_models, col_list)
Br_plot_mean
```

## Prostate
```{r}
Pr_ASV_RA <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/Prostate/ASV_RCV_RA_caret.RDS")
Pr_ASV_CLR <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/Prostate/ASV_RCV_CLR_caret.RDS")

Pr_Genus_RA <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/Prostate/Genus_RCV_RA_caret.RDS")
Pr_Genus_CLR <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/Prostate/Genus_RCV_CLR_caret.RDS")

Pr_models <- list("ASV CLR"=Pr_ASV_CLR, 
                  "ASV RA"=Pr_ASV_RA, 
                  "Genus CLR"=Pr_Genus_CLR, 
                  "Genus RA"=Pr_Genus_RA)

Pr_plot <- make_pROC_curve_mean(Pr_models, col_list)
Pr_plot
```

## Colon
```{r}
Cr_ASV_RA <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/CRC/ASV_RCV_RA_caret.RDS")
Cr_ASV_CLR <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/CRC/ASV_RCV_CLR_caret.RDS")

Cr_Genus_RA <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/CRC/Genus_RCV_RA_caret.RDS")
Cr_Genus_CLR <- readRDS("~/Private/CHAPTER_4/Prospec/ATP/Random_Forest/CRC/Genus_RCV_CLR_caret.RDS")


Cr_models <- list("ASV CLR"=Cr_ASV_CLR, 
                  "ASV RA"=Cr_ASV_RA, 
                  "Genus CLR"=Cr_Genus_CLR, 
                  "Genus RA"=Cr_Genus_RA)


Cr_plot_mean <- make_pROC_curve_mean(Cr_models, col_list)
Cr_plot_mean

saveRDS(Cr_plot_mean, "~/Private/CHAPTER_4/Publication_Markdowns/Display_items/intermediate_RDS2/Cr_plot_ATP_caret.RDS")
```

## Final Plot
```{r}
Ret_AUC_plot <- plot_grid(Br_plot_mean + ggtitle("Breast") + 
                            theme(legend.position = "none", plot.title = element_text(hjust=0.5)), 
                          Pr_plot + ggtitle("Prostate") + 
                            theme(legend.position = "none", plot.title = element_text(hjust=0.5)), 
                          # Cr_plot_mean + ggtitle("Colon") + 
                          #   theme(legend.position = "none", plot.title = element_text(hjust=0.5)), 
                          nrow=1, labels=c("A", "B"),
                          label_y=.1)

Ret_AUC_plot

Ret_AUC_plot_leg <- plot_grid(NULL,
                              Ret_AUC_plot, 
                              NULL,
                              get_legend(Br_plot_mean + theme(legend.position = "bottom")),
                              nrow=4,
                              rel_heights = c(-.3, 1,-.4, .2))
Ret_AUC_plot_leg

```


## Colon plot

```{r}
Cr_plot_PATH <- readRDS("~/Private/CHAPTER_4/Publication_Markdowns/Display_items/intermediate_RDS2/Cr_plot_PATH_caret.RDS")
Cr_plot_ATP <- readRDS("~/Private/CHAPTER_4/Publication_Markdowns/Display_items/intermediate_RDS2/Cr_plot_ATP_caret.RDS")


Cr_plot <- plot_grid(Cr_plot_PATH + ggtitle("Atlantic PATH. Pro. Colon") +
                       theme(legend.position = "none", plot.title = element_text(hjust=0.5)),
                     Cr_plot_ATP + ggtitle("ATP Pro. Colon") +
                       theme(legend.position = "none", plot.title = element_text(hjust=0.5)))

Cr_plot

# ggsave("~/Private/CHAPTER_4/Cleaned_Repo_D1/Display_Items/Main_Figures/Figure5.pdf",
#        width=13, height=7, units="in", dpi=600, plot=Cr_plot)
# 
# ggsave("~/Private/CHAPTER_4/Cleaned_Repo_D1/Display_Items/Main_Figures/Figure5.png",
#        width=13, height=7, units="in", dpi=600, plot=Cr_plot)
```

# top features

```{r}
remove_non_matching_caseIDS <- function(df){
  
    ## remove controls that don;t have associated case samples b/c they were removed due to poor sequencing
    df_cases <- df %>% filter(Case.Control=="case")
    CaseIDs <- df_cases$CaseID
    df_filt <- df %>% filter(CaseID %in% CaseIDs)

    ## remove cases that don't have any controls associtated because they were removed due to poor sequencing 
    ## these cases can only happen when the total number of samples with that CASEID is 1 
    Summ_data <- df_filt %>% group_by(CaseID) %>% summarize(n())
    remove_cases <- Summ_data$CaseID[which(Summ_data$`n()`==1)]
    
    if(length(remove_cases!=0)){
      message(paste0("removing ", remove_cases, " Case samples due to no controls"))
      test <- df_filt[-which(df_filt$CaseID %in% remove_cases),]
      return(test)  
    }

    
    return(df_filt)
  
}

match_age_diff <- function(df){
  
  case_df <- df %>% filter(Case.Control=="case")
  for(i in 1:nrow(df)){
    
    if(is.na(df[i,'age_diff']) & df[i,'Case.Control']=="control"){
      temp_caseID <- df[i,'CaseID']
      message(temp_caseID)
      temp_age_diff <- case_df$age_diff[which(case_df$CaseID==temp_caseID)]
      message(temp_age_diff)
      df[i,'age_diff'] <- temp_age_diff
    }
  }
  return(df)
}

library(readxl)
```


```{r}
base_metadata <- readRDS("~/Private/ATP/Metadata/Base_microbiome_metadata.RDS")
base_metadata$ParticipantKey <- as.character(base_metadata$ParticipantKey)
rownames(base_metadata) <- base_metadata$researcher



sheet_names <- c("ACR", "ACR_Treatment", "Core", "Core_Chronic", "HLQ", "Master_Data", "Case_Control_Match", "PANS", "PM_data", "UHLQ", "UHLQ_Cond")

Metadata <- list()

for(i in 1:length(sheet_names)){
  
  Metadata[[sheet_names[i]]] <- read_excel("~/Private/ATP/Metadata/ATP_MASTER_DATA.xls", sheet=i)
  
}


Metadata[[7]]$matchID <- seq(1:207)

case_control_match <- melt(Metadata[[7]][,-3], id.vars = 'matchID')



Shannon_div <- read.table("~/Private/ATP/Analysis_21_12_01/core-div-3000/raw_alpha/shannon.tsv",
                          sep="\t", header=T)
colnames(Shannon_div)[1] <- "researcher"

#faith pd
faith_pd <- read.table("~/Private/ATP/Analysis_21_12_01/core-div-3000/raw_alpha/faiths_pd.tsv",
                       sep="\t", header=T)
colnames(faith_pd)[1] <- "researcher"

#raw richness
richness <- read.table("~/Private/ATP/Analysis_21_12_01/core-div-3000/raw_alpha/observed_feats.tsv",
                       sep="\t", header=T)
colnames(richness)[1] <- "researcher"

#evenness
evenness <- read.table("~/Private/ATP/Analysis_21_12_01/core-div-3000/raw_alpha/evenness.tsv",
                       sep="\t", header=T)
colnames(evenness)[1] <- "researcher"


## 40 samples did not reach sequencing requirements.
Alpha_div <- Shannon_div %>% left_join(faith_pd, by="researcher") %>% left_join(richness, by="researcher") %>% left_join(evenness, by="researcher")


Full_data <- Alpha_div %>% inner_join(base_metadata, by="researcher")
Full_data$Case.Control <- factor(Full_data$ARM.x, levels=c("control", "case"))



Alpha_Breast <- Full_data %>% filter(Type=="Breast")
Alpha_Prostate <- Full_data %>% filter(Type=="Prostate")
Alpha_CRC <- Full_data %>% filter(Type=="Colon") 

Alpha_Breast$X <- Alpha_Breast$researcher
Alpha_Prostate$X <- Alpha_Prostate$researcher
Alpha_CRC$X <- Alpha_CRC$researcher

Alpha_Breast$CaseID <- case_control_match$matchID[match(Alpha_Breast$ParticipantKey, case_control_match$value)]
Alpha_Prostate$CaseID <- case_control_match$matchID[match(Alpha_Prostate$ParticipantKey, case_control_match$value)]
Alpha_CRC$CaseID <- case_control_match$matchID[match(Alpha_CRC$ParticipantKey, case_control_match$value)]

Alpha_Breast <- remove_non_matching_caseIDS(Alpha_Breast)
Alpha_Prostate <- remove_non_matching_caseIDS(Alpha_Prostate)
Alpha_CRC <- remove_non_matching_caseIDS(Alpha_CRC)

Breast_data_comb <- Alpha_Breast %>% select(faith_pd, shannon_entropy, pielou_evenness, observed_features, X, Case.Control, age_diff, Type, CaseID)

Prostate_data_comb <- Alpha_Prostate %>% select(faith_pd, shannon_entropy, pielou_evenness, observed_features, X, Case.Control, age_diff, Type, CaseID)

CRC_data_comb <- Alpha_CRC %>% select(faith_pd, shannon_entropy, pielou_evenness, observed_features, X, Case.Control, age_diff, Type, CaseID)

Combined_data <- rbind(Breast_data_comb, Prostate_data_comb, CRC_data_comb)


Breast_data <- Combined_data %>% filter(Type=="Breast")
CRC_data <- Combined_data %>% filter(Type=="Colon")
Prostate_data <- Combined_data %>% filter(Type=="Prostate")
```

```{r}

genera <- read.table("~/Private/ATP/Analysis_21_12_01/taxa_tabs/level6_tabs/feature-tab.tsv",
                     sep="\t", comment.char="", skip=1, header=T, check.names = F, row.names = 1)

ASV <- read.table("~/Private/ATP/Analysis_21_12_01/deblur_output/asv_table/feature-table.tsv",
                  sep="\t", comment.char = "", skip=1, header=T, check.names = F, row.names = 1)

genera_breast <- genera[,as.character(Breast_data$X)]
genera_prostate <- genera[,as.character(Prostate_data$X)]
genera_colon <- genera[,as.character(CRC_data$X)]

asv_breast <- ASV[,as.character(Breast_data$X)]
asv_prostate <- ASV[,as.character(Prostate_data$X)]
asv_colon <- ASV[,as.character(CRC_data$X)]
```

```{r}
ASV_Breast_CLR <- data.frame(apply(asv_breast + 1, 2, function(x){log(x) - mean(log(x))}), check.names = F)
ASV_CRC_CLR <- data.frame(apply(asv_colon + 1, 2, function(x){log(x) - mean(log(x))}), check.names = F)
ASV_Prostate_CLR <- data.frame(apply(asv_prostate + 1, 2, function(x){log(x) - mean (log(x))}), check.names = F)

ASV_Breast_CLR_flip <- data.frame(t(ASV_Breast_CLR), check.names = F)
ASV_Breast_CLR_flip$X <- rownames(ASV_Breast_CLR_flip)

ASV_CRC_CLR_flip <- data.frame(t(ASV_CRC_CLR), check.names = F)
ASV_CRC_CLR_flip$X <- rownames(ASV_CRC_CLR_flip)

ASV_Prostate_CLR_flip <- data.frame(t(ASV_Prostate_CLR), check.names = F)
ASV_Prostate_CLR_flip$X <- rownames(ASV_Prostate_CLR_flip)

Genus_Breast_CLR <- data.frame(apply(genera_breast + 1, 2, function(x){log(x) - mean(log(x))}), check.names = F)
Genus_CRC_CLR <- data.frame(apply(genera_colon + 1, 2, function(x){log(x) - mean(log(x))}), check.names = F)
Genus_Prostate_CLR <- data.frame(apply(genera_prostate + 1, 2, function(x){log(x) - mean(log(x))}), check.names = F)

Genus_Breast_CLR_flip <- data.frame(t(Genus_Breast_CLR), check.names = F)
Genus_Breast_CLR_flip$X <- rownames(Genus_Breast_CLR_flip)

Genus_CRC_CLR_flip <- data.frame(t(Genus_CRC_CLR), check.names = F)
Genus_CRC_CLR_flip$X <- rownames(Genus_CRC_CLR_flip)

Genus_Prostate_CLR_flip <- data.frame(t(Genus_Prostate_CLR), check.names = F)
Genus_Prostate_CLR_flip$X <- rownames(Genus_Prostate_CLR_flip)


ASV_Breast_RA <- sweep(asv_breast, 2, colSums(asv_breast), '/')
ASV_CRC_RA <- sweep(asv_colon, 2, colSums(asv_colon), '/')
ASV_Prostate_RA <- sweep(asv_prostate, 2, colSums(asv_prostate), '/')

ASV_Breast_RA_flip <- data.frame(t(ASV_Breast_RA), check.names=F)
ASV_Breast_RA_flip$X <- rownames(ASV_Breast_RA_flip)

ASV_CRC_RA_flip <- data.frame(t(ASV_CRC_RA), check.names = F)
ASV_CRC_RA_flip$X <- rownames(ASV_CRC_RA_flip)

ASV_Prostate_RA_flip <- data.frame(t(ASV_Prostate_RA), check.names = F)
ASV_Prostate_RA_flip$X <- rownames(ASV_Prostate_RA_flip)

Genus_Breast_RA <- sweep(genera_breast, 2, colSums(genera_breast), '/')
Genus_CRC_RA <- sweep(genera_colon, 2, colSums(genera_colon), '/')
Genus_Prostate_RA <- sweep(genera_prostate, 2, colSums(genera_prostate), '/')

Genus_Breast_RA_flip <- data.frame(t(Genus_Breast_RA), check.names = F)
Genus_Breast_RA_flip$X <- rownames(Genus_Breast_RA_flip)

Genus_CRC_RA_flip <- data.frame(t(Genus_CRC_RA), check.names = F)
Genus_CRC_RA_flip$X <- rownames(Genus_CRC_RA_flip)

Genus_Prostate_RA_flip <- data.frame(t(Genus_Prostate_RA), check.names = F)
Genus_Prostate_RA_flip$X <- rownames(Genus_Prostate_RA_flip)
```


```{r}
Comp_var_imp <- function(list_of_models, list_of_abs, metadata, top=5, taxa_classes){
  
  #make barplots of feat_imp
  barplots <- list()
  for(i in 1:length(list_of_models)){
    imp_df <- list_of_models[[i]]$finalModel$importance
    
    imp_df <- imp_df %>% data.frame() %>% arrange(desc(MeanDecreaseAccuracy)) %>% head(n=top)
    imp_df$feature <- rownames(imp_df)
    
    #dealing with ASVs
    if(i < 3){
      imp_df <- imp_df %>% left_join(Taxa_classifications, by="feature")
    }else{
      imp_df$Taxon <- gsub("[a-z]__.*(?=[a-z]__.*;[a-z]__.*)", "", imp_df$feature, perl=T)
    }
    
    #get SDs
    sd_df <- list_of_models[[i]]$finalModel$importanceSD %>% data.frame()
    colnames(sd_df) <- paste(colnames(sd_df), "_SD", sep="")
    sd_df$feature <- rownames(sd_df)
    
    imp_df <- imp_df %>% left_join(sd_df, by="feature")
    #imp_df$feature <- factor(imp_df$feature, levels=rev(imp_df$feature))
    imp_df$Taxon <- factor(imp_df$Taxon, levels=rev(imp_df$Taxon))
    

    p <- ggplot(imp_df, aes(y=MeanDecreaseAccuracy, x=Taxon)) +
      geom_col(fill="#999999", col="black") +
      geom_errorbar(aes(ymin=MeanDecreaseAccuracy-MeanDecreaseAccuracy_SD,
                        ymax=MeanDecreaseAccuracy+MeanDecreaseAccuracy_SD),
                    width=.2) +
      coord_flip()
      
    barplots[[i]] <- p
  }
  #make boxplots from ab data

  ab_boxplots <- list()
  for(i in 1:length(list_of_models)){
   
    
    ab_data <- list_of_abs[[i]][,as.character(barplots[[i]]$data$feature)]
    
    
    stopifnot(identical(rownames(ab_data), as.character(metadata$X)))
    
    ab_data$Case.Control <- metadata$Case.Control
    
    ab_data_melt <- melt(ab_data)

    if(i < 3){
      ab_data_melt$Taxon <- taxa_classes$Taxon[match(ab_data_melt$variable, taxa_classes$feature)]
    }else{
      ab_data_melt$Taxon <- gsub("[a-z]__.*(?=[a-z]__.*;[a-z]__.*)", "", ab_data_melt$variable, perl=T)
    }
    
    ab_data_melt$Taxon <- factor(ab_data_melt$Taxon, levels=rev(levels(barplots[[i]]$data$Taxon)))
  
    if(i %% 2==0){
      p <- ggplot(ab_data_melt, aes(x=Case.Control, y=log10(value+0.00001), fill=Case.Control)) +
        geom_boxplot(outlier.shape = NA) +
        geom_quasirandom(alpha=0.2) +
        coord_flip() +
        facet_wrap(~Taxon, nrow=top) +
        ylab("log10(Relative Abundance)") +
        xlab("")
    }else{
      p <- ggplot(ab_data_melt, aes(x=Case.Control, y=value, fill=Case.Control)) +
        geom_boxplot(outlier.shape = NA) +
        geom_quasirandom(alpha=0.2) +
        coord_flip() +
        facet_wrap(~Taxon, nrow=top) +
        ylab("Center-Log-Ratio Abundance") +
        xlab("")
    }

    
    ab_boxplots[[i]] <- p
  }
  
  return(list(barplots, ab_boxplots))
}


Taxa_classifications <- read.table("~/Private/ATP/Analysis_21_12_01/taxa/taxonomy/taxonomy.tsv", sep="\t",
                                   header=T, stringsAsFactors = F)

Taxa_classifications$Taxon <- gsub("[a-z]__.*(?=[a-z]__.*;[a-z]__.*)", 
                                   "", Taxa_classifications$Taxon, perl=T)
Taxa_classifications$Taxon <- make.unique(Taxa_classifications$Taxon)
Taxa_classifications$feature <- Taxa_classifications$Feature.ID
```

## Colon

```{r}
Cr_ab_data <- list(ASV_CRC_CLR_flip,
                   ASV_CRC_RA_flip,
                   Genus_CRC_CLR_flip,
                   Genus_CRC_RA_flip)

Cr_var_imp <- Comp_var_imp(list_of_models = Cr_models,
                           list_of_abs = Cr_ab_data,
                           metadata=CRC_data, top = 10, taxa_classes = Taxa_classifications)

fig6 <- plot_grid(Cr_var_imp[[1]][[3]],
          Cr_var_imp[[2]][[3]],
          nrow=1)

ggsave("~/Private/CHAPTER_4/Cleaned_Repo_D1/Display_Items/Main_Figures/Figure6.pdf",
       width=13, height=7, units="in", dpi=600, plot=fig6)
ggsave("~/Private/CHAPTER_4/Cleaned_Repo_D1/Display_Items/Main_Figures/Figure6.png",
       width=13, height = 7, units = "in", dpi=600, plot = fig6)

```
